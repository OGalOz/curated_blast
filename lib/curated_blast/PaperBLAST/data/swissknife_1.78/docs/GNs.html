<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>SWISS::GNs</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:operations@isb-sib.ch" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>

<ul>

	<li><a href="#name">Name</a></li>
	<li><a href="#description">Description</a></li>
	<li><a href="#inherits_from">Inherits from</a></li>
	<li><a href="#attributes">Attributes</a></li>
	<li><a href="#methods">Methods</a></li>
	<ul>

		<li><a href="#standard_methods">Standard methods</a></li>
		<li><a href="#reading_writing_methods">Reading/Writing methods</a></li>
	</ul>

	<li><a href="#transition">TRANSITION</a></li>
</ul>

<hr name="index" />
</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="name">Name</a></h1>
<p>SWISS::GNs.pm</p>
<p>
</p>
<hr />
<h1><a name="description">Description</a></h1>
<p><strong>SWISS::GNs</strong> represents the GN lines within an SWISS-PROT + TrEMBL
entry as specified in the user manual
 <a href="http://www.expasy.org/sprot/userman.html">http://www.expasy.org/sprot/userman.html</a> . The GNs object is a 
container object which holds a list of SWISS::GeneGroup objects.</p>
<p>
</p>
<hr />
<h1><a name="inherits_from">Inherits from</a></h1>
<p>SWISS::ListBase.pm</p>
<p>
</p>
<hr />
<h1><a name="attributes">Attributes</a></h1>
<dl>
<dt><strong><a name="list" class="item"><code>list</code></a></strong></dt>

<dd>
<pre>
  Each list element is a SWISS::GeneGroup object.</pre>
</dd>
<dt><strong><a name="and" class="item"><code>and</code> <em>(deprecated, for old format only)</em></a></strong></dt>

<dd>
<pre>
  Delimiter used between genes. Defaults to &quot; AND &quot;.</pre>
</dd>
<dt><strong><a name="or" class="item"><code>or</code> <em>(deprecated, for old format only)</em></a></strong></dt>

<dd>
<pre>
  Delimiter used between gene names. Defaults to &quot; OR &quot;.</pre>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="methods">Methods</a></h1>
<p>
</p>
<h2><a name="standard_methods">Standard methods</a></h2>
<dl>
<dt><strong><a name="new" class="item">new</a></strong></dt>

<dt><strong><a name="fromtext" class="item">fromText</a></strong></dt>

<dt><strong><a name="totext" class="item">toText</a></strong></dt>

</dl>
<p>
</p>
<h2><a name="reading_writing_methods">Reading/Writing methods</a></h2>
<dl>
<dt><strong><a name="text_newtext" class="item">text [($newText)]</a></strong></dt>

<dd>
<p>Sets the text of the GN line to the parameter if it is present, and returns
the (unwrapped) text of the line.  Also sets 'and' and 'or' delimiters to 
the first occurrences of the words &quot;OR&quot; and &quot;AND&quot; in the line, conserving
the case.</p>
</dd>
<dt><strong><a name="lowercase" class="item">lowercase <em>(deprecated, for old format only)</em></a></strong></dt>

<dd>
<p>Sets the GNs::and and GNs::or delimiters to their lower case
values.</p>
</dd>
<dt><strong><a name="uppercase" class="item">uppercase <em>(deprecated, for old format only)</em></a></strong></dt>

<dd>
<p>Sets the GNs::and and GNs::or delimiters to their upper case
values.</p>
</dd>
<dt><strong><a name="getfirst" class="item"><code>getFirst()</code></a></strong></dt>

<dd>
<p>Returns first gene name in gene line</p>
</dd>
<dt><strong><a name="gettags" class="item"><code>getTags($target)</code></a></strong></dt>

<dd>
<p>Returns evidence tags associated with $target</p>
<p>$target is a string</p>
</dd>
<dt><strong><a name="ispresent" class="item"><code>isPresent($target)</code></a></strong></dt>

<dd>
<p>Returns 1 if $target is present in the GN line</p>
<p>$target is a string</p>
</dd>
<dt><strong><a name="needsrecasing" class="item"><code>needsReCasing($target)</code></a></strong></dt>

<dd>
<p>If $target is present in the GN line, but wrongly cased, method returns the
matching name in its current case</p>
<p>$target is a string</p>
</dd>
<dt><strong><a name="replace" class="item">replace($newName, $target, $evidenceTag)</a></strong></dt>

<dd>
<p>Replaces the first GN object in the GN line whose text attribute is $target with
a new GN object whose text attribute is set to $newName and whose evidenceTags
attribute is is set using values set by splitting $evidenceTag on /, / (as name
is not being changed, programs should keep old tag and add new tag).  Does 
nothing if $target is not found.</p>
</dd>
<dt><strong><a name="delete" class="item"><code>delete($target)</code></a></strong></dt>

<dd>
<p>Removes synonym/single-member gene group matching $target. Note that if a &quot;Name&quot; 
is deleted, the first &quot;Synonym&quot; will be promoted to &quot;Name&quot;</p>
</dd>
<dt><strong><a name="addasnewsynonym" class="item">addAsNewSynonym($newName, $target, $evidenceTag, $location)</a></strong></dt>

<dd>
<p>Adds a new GN object (with text attribute set to new $newName, and evidenceTags
attribute set to ($evidenceTag)), as a synonym to the first gene group in which
$target is a gene name.  Does nothing if $target is not found.  Will not add a
duplicate gene name.  $location determines where in gene group new object is
added: if $location == 1, 2, 3, ..., new object added in the 1st, 2nd, 3rd, ... 
position; if $location == 0, new object added before $target; if $location
== -1, new object added after $target (default); if $location == -2, new object
added at end of gene group.  Note that if the new synonym is inserted in the 
first postion, it will become the &quot;Name&quot; and the previous &quot;Name&quot; will be downgraded
to first &quot;Synonym&quot;</p>
</dd>
<dt><strong><a name="addasnewgenegroup" class="item">addAsNewGeneGroup($newName, $target, $evidenceTag, $location)</a></strong></dt>

<dd>
<p>Adds a new GeneGroup object, comprising 1 GN object (with text attribute set to
new $newName, and evidenceTags attribute set to ($evidenceTag)). Will not add a
duplicate gene name.  $location and $target determine where in GNs line new 
group is added: if $location == 1, 2, 3, ..., new object added in the 1st, 2nd,
3rd, ... position; if $location == 0, new object added before $target; if
$location == -1, new object added after $target (default); if $location == -2,
new object added at end of GNs line.  Does nothing if $target is not found, and
$location == 0 or -1; otherwise $target does not need to be set.</p>
</dd>
<dt><strong><a name="replacegenegroup" class="item">replaceGeneGroup($newGeneGroup, $target)</a></strong></dt>

<dd>
<p>Replaces the first gene group containing $target with $newGeneGroup.  Creating
the $newGeneGroup correctly is the user's responsibility</p>
</dd>
<dt><strong><a name="getgenegroup" class="item"><code>getGeneGroup($target)</code></a></strong></dt>

<dd>
<p>Returns the first gene group that contains $target</p>
</dd>
<dt><strong><a name="settoor" class="item"><code>setToOr()</code></a></strong></dt>

<dd>
<p>Retruns a new GNs object, but with all GNs objects in a single gene group.  
Needed when adding 'C' to 'A and B', when the relationship of 'C' to 'A' and 
'B' is unknown: the universal use of ' or ' is the default delimeter for TrEMBL 
entries</p>
</dd>
</dl>
<p>
</p>
<hr />
<h1><a name="transition">TRANSITION</a></h1>
<p>The format of the GN line will change in 2004 from:</p>
<pre>
 GN   (CYSA1 OR CYSA OR RV3117 OR MT3199 OR MTCY164.27) AND (CYSA2 OR
 GN   RV0815C OR MT0837 OR MTV043.07C).

</pre>
<p>to:</p>
<pre>
 GN   Name=CysA1; Synonyms=CysA; OrderedLocusNames=Rv3117, MT3199;
 GN   ORFNames=MtCY164.27;
 GN   and
 GN   Name=CysA2; OrderedLocusNames=Rv0815c, MT0837; ORFNames=MTV043.07c;

</pre>
<p>This module supports both formats. To convert an entry from the old to
the new format, do:</p>
<pre>
 $entry-&gt;GNs-&gt;is_old_format(0);
</pre>

</body>

</html>
