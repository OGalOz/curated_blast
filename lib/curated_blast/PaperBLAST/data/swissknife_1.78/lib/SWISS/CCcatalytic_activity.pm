package SWISS::CCcatalytic_activity;

use vars qw($AUTOLOAD @ISA %fields);

use Carp;
use strict;
use SWISS::TextFunc;
use SWISS::BaseClass;
use SWISS::CC;


BEGIN {
    @ISA = ( 'SWISS::BaseClass' );

    %fields = (
        activity   => undef, # {  rhea => "", ec => "", reaction => "", participants => [], ev => "" }
        directions => undef, # [ { txt => "", rhea => "", ev => "" } ]
        form       => undef,
        _old       => undef  # [ [ txt, ev ], ]
    );
}


sub new {
    my $ref   = shift;
    my $class = ref($ref) || $ref;
    my $self  = new SWISS::BaseClass;
    $self->rebless($class);
    return $self;
}


sub fromText {
    my $class   = shift;
    my $textRef = shift;
    my $self    = new SWISS::CCcatalytic_activity;
    my $text    = $$textRef;
    $self->initialize();
    $text =~ s/ *-!- CATALYTIC ACTIVITY: +//;
    $text =~ s/ {2,}/ /g;

    if ( $text =~ /Reaction=/ ) { # new format (rhea/ec & reaction)
        my $core = {}; my $iscore = 0;
        foreach my $elem ( split /; /, $text ) {
            $elem =~ s/;$//;
            my ( $key, $val ) = ( "", "" );
            ( $key, $val ) = $elem =~ /^(\w+)=(.+)$/;
            if ( $key eq "Reaction" ) {
                $iscore = 1;
                $core->{reaction} = $val;
            }
            elsif ( $key eq "Xref" ) {
                my @ids = split /, /, $val;
                ( my $rhea_id = shift @ids ) =~ s/^Rhea://;
                if ( $iscore ) { # core (reaction)
                    $core->{rhea} = $rhea_id;
                    $core->{participants} = \@ids;
                }
                else           { # direction
                    my $dir = $self->{ directions }->[ -1 ] || { 'txt' => undef, 'ev' => undef };
                    $dir->{rhea} = $rhea_id;
                }
            }
            elsif ( $key eq "EC" ) {
                $core->{ec} = $val;
            }
            elsif ( $key eq "Evidence" ) {
                if ( $iscore ) { $core->{ev} = $val } # core (reaction)
                else           { # direction
                    my $dir = $self->{ directions }->[ -1 ] || { 'txt' => undef, 'rhea' => undef };
                    $dir->{ev} = $val;
                }
            }
            elsif ( $key eq "PhysiologicalDirection" ) {
                $iscore = 0;
                push @{ $self->{directions} }, { 'txt' => $val, 'rhea' => undef, 'ev' => undef };
            }
        }
        $self->{activity} = $core;
    }
    else { # old unstructured format
        my $has_new_style_ev = $text =~ /ECO:/;
        $self->{_old} = SWISS::CC::parse2Blocks( $text );
        # set evidenceTags 4 compatibility with existing ...EvidenceTag methods in BaseClass! + old tests (fixme: clean/remove that mess!?)
        # but the real evidences are now within blocks (2nd field of block array)
        my @evs = map { split /, ?/, $_->[1] } grep { $_->[1] } @{ $self->{_old} };
        my $ev  = $has_new_style_ev ? " {" . join( ", ", @evs ) . "}" : "{" . join( ",", @evs ) . "}";
        $self->evidenceTags( $ev ) if @evs;
    }

    $self->{ _dirty } = 0;
    return $self;
}


sub topic {
    return "CATALYTIC ACTIVITY";
}


sub activity_rhea { # get (or set) rhea (id without "Rhea:" prefix, but with RHEA:) e.g. (RHEA:18037)
    my ( $self, $value, $ev ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if ( defined $value ) {
            my $new_ev = $ev ? $ev : $self->{ activity }->{ev};
            $self->{activity}->{rhea} = $value;
            $self->{activity}->{ev}   = $new_ev;
        }

        return $self->{activity} ? $self->{activity}->{rhea} : "";
    }
    else { return "" }
}

sub activity_reaction { # get (or set) reaction
    my ( $self, $value, $ev_only_for_old_format ) = @_;

    if ( !$self->{_old} ) { # new (rhea/ec + reaction) format
        if ( defined $value ) {
            $self->{activity}->{reaction} = $value;
        }

        return $self->{activity} ? $self->{activity}->{reaction} : "";
    }
    else { # old format
        if ( defined $value ) {
            my $new_ev = $ev_only_for_old_format ? $ev_only_for_old_format : " {" . join( ", ", map { split /, ?/, $_->[1] } grep { $_->[1] } @{ $self->{_old} } ) . "}";
            $self->evidenceTags( $new_ev );
            $self->{_old} = [ [ $value, $new_ev ] ];
        } # -> ! not tested
        return join( ". ", map { $_->[0] } @{ $self->{_old} } )."."
    }
}

sub activity_ec { # get (or set) ec
    my ( $self, $value, $ev ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if ( defined $value ) {
            my $new_ev = $ev ? $ev : $self->{ activity }->{ev};
            $self->{activity}->{ec} = $value;
            $self->{activity}->{ev} = $new_ev;
        }

        return $self->{activity} ? $self->{activity}->{ec} : "";
    }
    else { return "" }
}

sub activity_participant {
    my ( $self, $values ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if (defined $values) {
            $self->{activity}->{participants} = $values;
        }
        return $self->{activity} ? $self->{activity}->{participants} : [];
    }
    else { return [] }
}

sub directions { # get (or set) direction(s) structures
    my ( $self, $values ) = @_;

    if ( !$self->{_old} ) { # only works with new (rhea/ec + reaction) format
        if ( defined $values && ref $values eq "ARRAY") {
            $self->{directions} = $values;
        }
        return $self->{directions};
    }
    else { return [] }
}


sub toString {
    my ( $self ) = @_;

    my $activity_txt = $self->_activity_to_text();

    my $text = "CC   -!- CATALYTIC ACTIVITY:";
    if ( !$self->{_old} ) { # new (rhea/ec, reaction) format
        #my $form = $self->{form} ? " $self->{form}:" : "";
        $text .= "\n";
        $text .= SWISS::TextFunc->wrapOn( 'CC       ',"CC         ", $SWISS::TextFunc::lineLength, $activity_txt );
        map {
            my $direction_txt = _direction_struct_to_text( $_ );
            $text .= SWISS::TextFunc->wrapOn( 'CC       ',"CC         ", $SWISS::TextFunc::lineLength, $direction_txt );
        } @{ $self->{directions} };
    }
    else { # old unstructured format
        $text .= " ".$activity_txt;
        $text = SWISS::TextFunc->wrapOn('',"CC       ", $SWISS::TextFunc::lineLength, $text );
    }

    return $text;
}

sub _activity_to_text {
    my ( $self, $no_ev ) = @_;
    if ( !$self->{_old} ) { # new structured format
        my $activity     = $self->{ activity };
        my $activity_txt = "Reaction=$activity->{reaction};";
        if ( $activity->{rhea} ) {
            $activity_txt .= " Xref=Rhea:$activity->{rhea}";
            if ( $activity->{participants} )  { $activity_txt .= ", " . join( ", ", @{ $activity->{participants} } ) }
            $activity_txt .= ";";
        }
        if ( $activity->{ec} )            { $activity_txt .= " EC=$activity->{ec};" }
        if ( $activity->{ev} && !$no_ev ) { $activity_txt .= " Evidence=$activity->{ev};" }
        return $activity_txt;
    }
    else { # old format
        return ( $no_ev ? join( ". ", map { $_->[0] } @{ $self->{_old} } )."." : SWISS::CC::blocks2String( $self->{_old}, $self->evidenceTags )."." );
    }
}
sub _direction_struct_to_text {
    my ( $direction, $no_ev ) = @_;
    my $direction_txt = "";
    if ( $direction->{txt} )           { $direction_txt .= "PhysiologicalDirection=$direction->{txt};" }
    if ( $direction->{rhea} )          { $direction_txt .= " Xref=Rhea:$direction->{rhea};" }
    if ( $direction->{ev} && !$no_ev ) { $direction_txt .= " Evidence=$direction->{ev};" }
    return $direction_txt;
}

sub comment { # unwrapped content (without ev)
    my ($self, $with_ev) = @_;

    my $activity_txt  = $self->_activity_to_text (  "no_ev=true" );
    my $direction_txt = "";
    map { $direction_txt .= ( $direction_txt ? " " : "" ) . _direction_struct_to_text( $_ , "no_ev=true" ); } @{ $self->{directions} };
    return $activity_txt . ( $direction_txt ? " $direction_txt" : "" );
}


#sub sort {
#    my $self = shift;
#    $self->{ list } = sort { lc( $a->[0] ) cmp lc( $b->[0] ) } @{ $self->{ list } };
#}


1;

__END__

=head1 Name

SWISS::CCcatalytic_activity

=head1 Description

B<SWISS::CCcofactor> represents a comment on the topic 'CATALYTIC ACTIVITY'
within a Swiss-Prot or TrEMBL entry as specified in the user manual
http://www.expasy.org/sprot/userman.html .  Comments on other topics are stored
in other types of objects, such as SWISS::CC (see SWISS::CCs for more information).

Collectively, comments of all types are stored within a SWISS::CCs container
object.

=head1 Inherits from

SWISS::BaseClass

=head1 Attributes

=over

=item topic

The topic of this comment ('CATALYTIC ACTIVITY').

=item comment

The "text" version of this comment (without evidences and new lines).


=back
=head1 Methods

=item activity_rhea

The rhea id (e.g. RHEA:18037) of the reaction (only for new structured CC CATALYTIC ACTIVITY)

=item activity_rhea( $new_rhea_id, $new_optional_reaction_ev )

Set rhea activity to $new_rhea_id (and reaction ev to $new_optional_reaction_ev if provided)
(only for new structured CC CATALYTIC ACTIVITY)

=item activity_reaction

The reaction txt (e.g. agmatine + H2O = N-carbamoylputrescine + NH4(+))

=item activity_reaction( $new_reaction_txt )

Set reaction txt to $new_reaction_txt
(on old unstructured format: extra $ev could be added)

=item activity_ec

The reaction EC (e.g. 3.5.3.12)

=item activity_ec( $new_reaction_ec, , $new_optional_reaction_ev )

Set reaction EC to $new_reaction_ec (and reaction ev to $new_optional_reaction_ev if provided)
(only for new structured CC CATALYTIC ACTIVITY)

=item activity_participant

The reaction participants as a String Array (e.g. [ "ChEBI:CHEBI:15377", "ChEBI:CHEBI:28938", "ChEBI:CHEBI:58145", "ChEBI:CHEBI:58318" ])
(participant are shown as dbtype:id e.g. in "ChEBI:CHEBI:15377" CheBI is db type, CHEBI:15377 is (CheBi) id)
(only for new structured CC CATALYTIC ACTIVITY)

=item activity_participant( $new_participant_array )

Set reaction participants to $new_participant_array (array of strings)
(only for new structured CC CATALYTIC ACTIVITY)


=item directions

The directions: Array of direction structures as hash e.g. [ { 'txt' => 'left-to-right', 'rhea' => 'RHEA:16846', 'ev' => '{ECO:0000269|PubMed:29420286}' } ]
p.s. in well formed UniProtKB data 'txt' will be 'right-to-left' or 'left-to-right' (in that order); there should be no more than 2 directions (minimum: 0)

=item direction( $new_directions )

Set directions Array to $new_directions ...
(only for new structured CC CATALYTIC ACTIVITY)

=head2 Standard methods

=over

=item new

=item fromText

=item toString

Returns a string representation of this comment.

=back
