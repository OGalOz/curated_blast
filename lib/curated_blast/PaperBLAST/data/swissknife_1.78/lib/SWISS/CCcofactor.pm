package SWISS::CCcofactor;

use vars qw($AUTOLOAD @ISA %fields);

use Carp;
use strict;
use SWISS::TextFunc;
use SWISS::ListBase;
use SWISS::CC;


BEGIN {
  @ISA = ( 'SWISS::ListBase' );
  
  %fields = (
          form        => undef,  
	      note        => undef, # [ txt, ev ]
	      note_blocks => undef # [ [txt, ev ] ]
	    );
}


sub new {
  my $ref = shift;
  my $class = ref($ref) || $ref;
  my $self = new SWISS::ListBase;
  $self->rebless($class);
  return $self;
}


sub fromText {
    my $class = shift;
    my $textRef = shift;
    my $self = new SWISS::CCcofactor;
    my $text = $$textRef;
    $self->initialize();
  
    $text =~ s/ *-!- COFACTOR: +//;
    $text =~ s/; {2,}/; /g;
    $text =~ s/, {2,}/, /g;
    
    if ( $text !~ /Name=/ && $text !~ /Note=/ ) { # old format!
    	  my $note = $text;
    	( my $note_ev = $1 ) if $note =~ s/($SWISS::TextFunc::evidencePattern)\.?//;
    	$self->{ note } = [ $note.".", $note_ev ] if $note;
    }
    else { # new format
        ( my $form    = $1 )           if $text =~ s/^([^:]+): (?=N)//;
        ( my $note    = $1 ) =~ s/;$// if $text =~ s/ ?Note=(.+)$//;
        $self->{ form }  = $form if $form;
        $self->{ note_blocks } = SWISS::CC::parse2Blocks( $note ) if $note; #[ $note, $note_ev ] if $note;
        foreach my $name ( split / +(?=Name=)/, $text ) {
            my $ev = $name =~ s/ +Evidence=($SWISS::TextFunc::evidencePattern);// ? " ".$1 : undef;
            $self->add( [ $name, $ev ] );
        } # $name = e.g. "Name=[2Fe-2S] cluster; Xref=ChEBI:CHEBI:49601;" includes chebi xref! not further parsed!
    }

    $self->{_dirty} = 0;
    return $self;
}


sub topic {
	return "COFACTOR";
}


sub toString {
    my ( $self ) = @_;
    
    my $form = $self->{form} ? " $self->{form}:" : "";
    my $text = "CC   -!- COFACTOR:$form\n";
    
    foreach my $name_ev ( @{ $self->{list} } ) {
        my ( $name, $ev ) = @$name_ev;
        my   $line = $name;
        if ( $ev ) { $ev=~s/^ +//; $line .= " Evidence=".$ev.";" }
        $text .= SWISS::TextFunc->wrapOn( 'CC       ',"CC         ", $SWISS::TextFunc::lineLength, $line, "(?<=;) ", "(?<=,) ", $SWISS::TextFunc::textWrapPattern1 );
    }
    if ( $self->{note} ) { # old format
     	my ( $note, $note_ev ) = @{ $self->{ note } };
        $note_ev ||= "";
        $note      = "Note=" . $note . $note_ev . ";";
        $text     .= SWISS::TextFunc->wrapOn( 'CC       ',"CC       ", $SWISS::TextFunc::lineLength, $note );
    }
    if ( $self->{note_blocks} ) {
        my $note  = "Note=" . SWISS::CC::blocks2String( $self->{ note_blocks }, "" ) . ";";
        $text    .= SWISS::TextFunc->wrapOn( 'CC       ',"CC       ", $SWISS::TextFunc::lineLength, $note );
    }
    
    return $text; 
}


sub comment {
    my ( $self, $with_ev ) = @_;
    
    my $text = "";
    foreach my $name_ev ( @{ $self->{list} } ) {
        my ( $name, $ev ) = @$name_ev;
        $text .= ( $text ? " " : "" ) . $name;
        if ( $ev && $with_ev ) { $ev=~s/^ +//; $text .= " Evidence=".$ev.";" }
    }
    if ( $self->{ note } ) {
    	my ( $note, $note_ev ) = @{ $self->{ note } };
    	$note_ev = "" unless $with_ev && $note_ev;
        $note      = "Note=" . $note . $note_ev . ";";
        $text .= ( $text ? " " : "" ) . $note;
    }
    
    return $text;
}


sub structured_names {
    my ( $self, $values ) = @_;

    if ( defined $values ) { # set
        $self->{list} = \ map { [ "Name=".$_->{name}."; Xref=ChEBI:".$_->{chebi}, $_->{ev} ] } @{ $values };
    }
    my @out = map { # get
        my $elem = {};
        $elem->{ev} = $_->[1];                                    # e.g. ECO:0000255|HAMAP-Rule:MF_01841
        my @vals = grep { $_ ne '' } split /Name=|Xref=/, $_->[0];
        ( $elem->{name}  = $vals[0] || "" ) =~ s/; ?$//;          # e.g. FAD
        ( $elem->{chebi} = $vals[1] || "" ) =~ s/^ChEBI:|; ?$//g; # e.g. CHEBI:57692
        print "----STRUCT COF-->[".$elem->{name}."]\n";
        $elem;
    } @{ $self->{list} };
    return \@out;
} # returns Array of { name->..., chebi->..., ev->... }


#sub sort {
#    my $self = shift;
#    $self->{ list } = sort { lc( $a->[0] ) cmp lc( $b->[0] ) } @{ $self->{ list } };
#}


1;

__END__

=head1 Name

SWISS::CCcofactor

=head1 Description

B<SWISS::CCcofactor> represents a comment on the topic 'COFACTOR'
within a Swiss-Prot or TrEMBL entry as specified in the user manual
http://www.expasy.org/sprot/userman.html .  Comments on other topics are stored
in other types of objects, such as SWISS::CC (see SWISS::CCs for more information).

Collectively, comments of all types are stored within a SWISS::CCs container
object.

=head1 Inherits from

SWISS::ListBase.pm

=head1 Attributes

=over

=item topic

The topic of this comment ('COFACTOR').

=item comment

The "text" version of this comment (without evidences and new lines).

=item note

The note and evidence (Note= in new format or full description in old format)
reference to an array of [ $note, $note_ev ] (strings)

=item elements

An array of [name_str, evidence_tags_str], if any.

=back
=head1 Methods

=item structured_names

(only for new structured CC COFACTOR)
The names as array of 'name' txt, 'chebi' id, 'ev' string (hash) structures
e.g. [ { 'name'->'FAD', 'chebi'->'CHEBI:57692', 'ev'->'ECO:0000250' } ]

=item structured_names( $new_values )

(only for new structured CC COFACTOR)
Set names with array of 'name' txt, 'chebi' id, 'ev' string (hash) structures like returned by get-er version


=head2 Standard methods

=over

=item new

=item fromText

=item toString

Returns a string representation of this comment.

=back
