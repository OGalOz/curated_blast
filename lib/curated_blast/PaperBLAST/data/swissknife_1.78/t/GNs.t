# Swissknife Test Harness Script for GNs
# 
# Purpose: 
# Check whether the advanced parser in GNs works well.

# * Test loading
BEGIN { 
  $| = 1; print "1..2\n"; 
  use vars qw($loaded); 
  $^W = 0;
}

END {print "not ok 1\n" unless $loaded;}


$loaded = 1;
print "ok 1\n";    # 1st test passes.

sub test ($$;$) {
    my($num, $true,$msg) = @_;
    print($true ? "ok $num\n" : "not ok $num $msg\n");
}

use SWISS::GNs;
my $where = -d 't' ? "t/" : "";
my $testout  = "${where}GNs.txl.out";
my $expectedout = "${where}GNs.txl.expected";

open (OUT, ">$testout");

my $out="";
my $text = "GN   (A{EA1} OR B{EI2}) And C.\n";
my $gn=SWISS::GNs->fromText(\$text,1);
$gn->toText(\$out);
print OUT $out;

#ev tags
for my $gg ($gn->elements) {
 for(my $i=0;$i<$gg->size;$i++) {
  ${$gg->list}[$i]->evidenceTags($1.($2+1).$3) if
${$gg->list}[$i]->evidenceTags =~ /^(.*)(\d+)(.*)$/;
;
 }
}
$out="";
$gn->toText(\$out);
print OUT $out;


#delimiter
$gn->lowercase;
$out="";
$gn->toText(\$out);
print OUT $out;

my $gn2=SWISS::GNs->fromText(\"GN   A{EA1} or B{EA2}.\n",1);
$out="";
$gn2->toText(\$out);
print OUT $out;

$gn2=SWISS::GNs->fromText(\"GN   Timeo danaos, and dona ferentes.\n",1);
$gn2->and(" et ");
$out="";
$gn2->toText(\$out);
print OUT $out;

#remove GN
$gn->head->pop;
$out="";
$gn->toText(\$out);
print OUT $out;


$gn->head->set;
$out="";
$gn->toText(\$out);
print OUT $out;


$gn->head->set;
$out="";
$gn->toText(\$out);
print OUT $out || "(Nothing)\n";


#remove genegroups
$gn=SWISS::GNs->fromText(\$text,1);
@{$gn->list}=();
$out="";
$gn->toText(\$out);
print OUT $out || "(Nothing)\n";

#adding genegroups
@{$gn->list} = SWISS::GeneGroup->fromText("Wallace OR Gromit");
$out="";
$gn->toText(\$out);
print OUT $out;


#wrapping
push @{$gn->list}, SWISS::GeneGroup->fromText("So long vict OR ious, happy 'n gl OR ious, long to reign over us");
$out="";
$gn->toText(\$out);
print OUT $out;

unshift @{$gn->list}, SWISS::GeneGroup->fromText("Foobar");
$out="";
$gn->toText(\$out);
print OUT $out;

$gn->is_old_format(0);
$out="";
$gn->toText(\$out);
print OUT $out;

$out = ' ' . $out;
$gn = SWISS::GNs->fromText(\$out);
$gn->item(2)->Names->splice (1,1);
$out="";
$gn->toText(\$out);
print OUT $out;


close OUT;
print "checking expected output...\n";

test 2, system('diff', $testout, $expectedout) == 0, "diff $testout $expectedout";





