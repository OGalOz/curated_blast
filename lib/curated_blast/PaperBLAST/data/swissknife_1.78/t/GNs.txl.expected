GN   (A{EA1} OR B{EI2}) And C.
GN   (A{EA2} OR B{EI3}) And C.
GN   (A{EA2} or B{EI3}) and C.
GN   A{EA1} or B{EA2}.
GN   Timeo danaos, et dona ferentes.
GN   A{EA2} and C.
GN   C.
(Nothing)
(Nothing)
GN   Wallace OR Gromit.
GN   (Wallace OR Gromit) And
GN   (So long vict OR ious, happy 'n gl OR ious, long to reign over us).
GN   Foobar And (Wallace OR Gromit) And
GN   (So long vict OR ious, happy 'n gl OR ious, long to reign over us).
GN   Name=Foobar;
GN   and
GN   Name=Wallace; Synonyms=Gromit;
GN   and
GN   Name=So long vict;
GN   Synonyms=ious, happy 'n gl, ious, long to reign over us;
 GN   Name=Foobar;
 GN   and
 GN   Name=Wallace; Synonyms=Gromit;
 GN   and
 GN   Name=So long vict; Synonyms=happy 'n gl, ious, long to reign over us;
